### PR TITLE
[ROCm][CI] Fix ROCm LoRA Transformers fallback with full CUDA graphs

### PR DESCRIPTION
**Pull request by AndreasKaratzas**
Fixes `LoRA %N`

- Restrict the gfx950 `wvSplitKrc` skinny GEMM path to 2D activations.
- Make the vLLM C RMSNorm provider flatten N-D activations by token row and reshape back.
- Initialize torch-compile support on the actual HF decoder instance for dynamic remote-code classes.

The original failure was `Unsupported N value: 512,2048,1` in `wvSplitKrc`. The selector counted flattened tokens from a 3D Transformers activation, but `wvSplitKrc` reads `size(0)` and is only correct for 2D row-major inputs. After fixing that selector bug, full decode graph capture exposed the same N-D activation mismatch in the `vllm_c` RMSNorm provider. Flattening RMSNorm inputs matches its row-wise kernel contract and preserves the public tensor shape.

- Known-good commit: `0e39202ca911319c7747a2f9d5a0c162fdff4fd9`.
- Failing commit observed in CI: `7845379230c48c9103ab71682ee05977dfbbdfce`.
- The likely trigger is the update to transformers v5

cc @kenroche

### CODE REVIEW

**Comment by hmellor** [comment]
The third dimension when using the Transformers modelling backend is always empty. Wouldn't it be better if ROCm ignored it in the same way CUDA does?

File: vllm/model_executor/models/transformers/base.py [hunk 1]
```diff
@@ -169,6 +172,7 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
         # Init on "meta" to delay allocating GPU tensors
         with init_on_device_without_buffers("meta"):
             self.model: PreTrainedModel = AutoModel.from_config(**from_config_kwargs)
+        self._initialize_decoder_torch_compile()
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/base.py]: Wouldn't this double wrap the class since `support_torch_compile` is applied before `self.model` is instantiated?

**Comment by AndreasKaratzas** [comment]
> The third dimension when using the Transformers modelling backend is always empty. Wouldn't it be better if ROCm ignored it in the same way CUDA does?

I understand the motivation for parity, but this ROCm path is a specialized custom kernel rather than a general `F.linear` equivalent. The selector computes logical `N` as `x.numel() // x.size(-1)`, but `wvSplitKrc` itself reads `in_a.size(0)`. So unless we explicitly flatten before calling it, the kernel sees the wrong `N` for Transformers-shaped activations. N-D support for `wvSplitKrc` (if found more benefitial) would be added separately in any case.

File: vllm/model_executor/models/transformers/base.py (hunk 1 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/models/transformers/base.py]: In the normal path this should be a no-op rather than a double wrap. If `AutoModel.from_config()` instantiates the class patched by `_decorate_for_torch_compile()`, then `TorchCompileWithNoGuardsWrapper` is already in the class bases and the wrapped `__init__` has already set `do_not_compile` on the instance. `initialize_torch_compile()` checks both of those and returns early. The helper is for the remote-code/dynamic-class case where the actual decoder instance returned by `from_config()` is not the same class we patched beforehand.

**Review by hmellor — CHANGES_REQUESTED** [review]
IIUC this PR doesn't account for the compoilation of the vision encoder

File: vllm/kernels/vllm_c.py [hunk 2]
```diff
@@ -28,14 +27,23 @@ def rms_norm(
         # Kernel requires weight tensor, pass ones
         weight = torch.ones(x.shape[-1], device=x.device, dtype=x.dtype)
     assert variance_size is None
+    # NOTE: This check is necessary for ROCm. However, CUDA (A100) was also
+    # found to score better performance with this path, so we use it for all
+    # CUDA-alike platforms.
```
Comment (hmellor) [inline vllm/kernels/vllm_c.py]: I would also add that this is only for the Transformers modelling backend (the only case where x will be 3D)

File: tests/kernels/ir/test_layernorm.py
Comment (hmellor) [inline tests/kernels/ir/test_layernorm.py:1]: The 0-th dimension of `x` should be `1` as that is what it will be in the Transformers modelling backend

Comment (AndreasKaratzas) [reply tests/kernels/ir/test_layernorm.py:1]: Done :)

File: vllm/kernels/vllm_c.py (hunk 2 above)
Comment (AndreasKaratzas) [reply vllm/kernels/vllm_c.py]: Done :)

**Review by hmellor — APPROVED** [review]

File: vllm/kernels/vllm_c.py [hunk 3]
```diff
@@ -28,14 +29,23 @@ def rms_norm(
...
     assert variance_size is None
+    # ROCm's vLLM C RMSNorm kernel operates on contiguous 2D tensors.
+    # Higher-rank callers still normalize over the last dimension, so flatten
+    # all leading dims. reshape handles strided views from q/k/v splits.
+    if IS_ROCM and (x.dim() > 2 or not x.is_contiguous()):
+        original_shape = x.shape
+        x = x.reshape(-1, original_shape[-1])
+        output = torch.empty_like(x)
+        torch.ops._C.rms_norm(output, x, weight, epsilon)
+        return output.reshape(original_shape)
```
Comment (hmellor) [inline vllm/kernels/vllm_c.py:40]: Would it be more performant to only reshape in the non-contiguous case?

Comment (AndreasKaratzas) [reply vllm/kernels/vllm_c.py:40]: I don't think that would be more performant in the contiguous N-D case. For contiguous tensors, `reshape(-1, hidden)` is metadata-only and does not copy, so it just lets us call the rank-2 RMSNorm path. The expensive case is the non-contiguous Transformers split/view case, where `reshape` may materialize a contiguous 2D tensor, but that is the case we need to normalize before calling this. So I would keep this as-is.

Comment (hmellor) [reply vllm/kernels/vllm_c.py:40]: Oh excellent, I was under the impression that reshape would always copy

**Review by hmellor — APPROVED** [review]
LGTM
